Wake On Lan DNS errors when not supplying an optional keyword / parameter

<!-- READ THIS FIRST:
  - If you need additional help with this template, please refer to https://www.home-assistant.io/help/reporting_issues/
  - Make sure you are running the latest version of Home Assistant before reporting an issue: https://github.com/home-assistant/core/releases
  - Do not report issues for integrations if you are using custom components or integrations.
  - Provide as many details as possible. Paste logs, configuration samples and code into the backticks.
  DO NOT DELETE ANY TEXT from this template! Otherwise, your issue may be closed without comment.
-->
## The problem
<!-- 
  Describe the issue you are experiencing here to communicate to the
  maintainers. Tell us what you were trying to do and what happened.
-->

The WoL integration causes issues with CoreDNS when the "Host" parameter is not set. The documentation suggests that "HOST" is an optional key however, whenever it is not set the CoreDNS container reports the following error messages:

Oct 06 14:03:24 HomeAssistant 1008d0f60de7[458]: [ERROR] plugin/errors: 2 none.local.hass.io. A: unreachable backend: no upstream host
Oct 06 14:03:27 HomeAssistant 1008d0f60de7[458]: [INFO] 172.30.32.1:60123 - 49608 "A IN none.local.hass.io. udp 36 false 512" NOERROR - 0 0.00039084s

Port number changes but IP remains static.

https://imgur.com/a/PJ0Exz1
https://imgur.com/a/jwNewy3

## Environment
<!--
  Provide details about the versions you are using, which helps us to reproduce
  and find the issue quicker. Version information is found in the
  Home Assistant frontend: Configuration -> Info.
-->

- Home Assistant Core release with the issue: 0.115.6
- Last working Home Assistant Core release (if known): 0.114.4
- Operating environment (OS/Container/Supervised/Core):  Supervisor
- Integration causing this issue: Wake On Lan
- Link to integration documentation on our website: https://www.home-assistant.io/integrations/wake_on_lan/

## Problem-relevant `configuration.yaml`
<!--
  An example configuration that caused the problem for you. Fill this out even
  if it seems unimportant to you. Please be sure to remove personal information
  like passwords, private URLs and other credentials.
-->

```yaml
See attached screenshot
```

## Traceback/Error logs
<!--
  If you come across any trace or error logs, please provide them.
-->

```txt
See attached screenshot
```

## Additional information
When "Host" keyword is supplied in the YAML then the issues goes away however, the documentation states this is optional.

--- a/CODEOWNERS
+++ b/CODEOWNERS
@@ -521,6 +521,7 @@ homeassistant/components/vizio/* @raman325
 homeassistant/components/vlc_telnet/* @rodripf @dmcc
 homeassistant/components/volkszaehler/* @fabaff
 homeassistant/components/volumio/* @OnFreund
+homeassistant/components/wake_on_lan/* @ntilley905
 homeassistant/components/waqi/* @andrey-git
 homeassistant/components/watson_tts/* @rutkai
 homeassistant/components/weather/* @fabaff

--- a/CODEOWNERS
+++ b/CODEOWNERS
@@ -521,7 +521,6 @@ homeassistant/components/vizio/* @raman325
 homeassistant/components/vlc_telnet/* @rodripf @dmcc
 homeassistant/components/volkszaehler/* @fabaff
 homeassistant/components/volumio/* @OnFreund
-homeassistant/components/wake_on_lan/* @ntilley905
 homeassistant/components/waqi/* @andrey-git
 homeassistant/components/watson_tts/* @rutkai
 homeassistant/components/weather/* @fabaff

--- a/homeassistant/components/wake_on_lan/manifest.json
+++ b/homeassistant/components/wake_on_lan/manifest.json
@@ -3,5 +3,5 @@
   "name": "Wake on LAN",
   "documentation": "https://www.home-assistant.io/integrations/wake_on_lan",
   "requirements": ["wakeonlan==1.1.6"],
-  "codeowners": []
+  "codeowners": ["@ntilley905"]
 }

--- a/homeassistant/components/wake_on_lan/switch.py
+++ b/homeassistant/components/wake_on_lan/switch.py
@@ -86,6 +86,10 @@ class WolSwitch(SwitchEntity):
             Script(hass, off_action, name, domain) if off_action else None
         )
         self._state = False
+        self._is_dummy = False
+
+        if self._host is None:
+            self._is_dummy = True
 
     @property
     def is_on(self):
@@ -114,31 +118,38 @@ class WolSwitch(SwitchEntity):
 
         wakeonlan.send_magic_packet(self._mac_address, **service_kwargs)
 
+        if self._is_dummy:
+            self._state = True
+
     def turn_off(self, **kwargs):
         """Turn the device off if an off action is present."""
         if self._off_script is not None:
             self._off_script.run(context=self._context)
 
+        if self._is_dummy:
+            self._state = False
+
     def update(self):
         """Check if device is on and update the state."""
-        if platform.system().lower() == "windows":
-            ping_cmd = [
-                "ping",
-                "-n",
-                "1",
-                "-w",
-                str(DEFAULT_PING_TIMEOUT * 1000),
-                str(self._host),
-            ]
-        else:
-            ping_cmd = [
-                "ping",
-                "-c",
-                "1",
-                "-W",
-                str(DEFAULT_PING_TIMEOUT),
-                str(self._host),
-            ]
+        if not self._is_dummy:
+            if platform.system().lower() == "windows":
+                ping_cmd = [
+                    "ping",
+                    "-n",
+                    "1",
+                    "-w",
+                    str(DEFAULT_PING_TIMEOUT * 1000),
+                    str(self._host),
+                ]
+            else:
+                ping_cmd = [
+                    "ping",
+                    "-c",
+                    "1",
+                    "-W",
+                    str(DEFAULT_PING_TIMEOUT),
+                    str(self._host),
+                ]
 
-        status = sp.call(ping_cmd, stdout=sp.DEVNULL, stderr=sp.DEVNULL)
-        self._state = not bool(status)
+            status = sp.call(ping_cmd, stdout=sp.DEVNULL, stderr=sp.DEVNULL)
+            self._state = not bool(status)

--- a/homeassistant/components/wake_on_lan/switch.py
+++ b/homeassistant/components/wake_on_lan/switch.py
@@ -57,7 +57,7 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
                 broadcast_port,
             )
         ],
-        True,
+        host is not None,
     )
 
 
@@ -86,10 +86,7 @@ class WolSwitch(SwitchEntity):
             Script(hass, off_action, name, domain) if off_action else None
         )
         self._state = False
-        self._is_dummy = False
-
-        if self._host is None:
-            self._is_dummy = True
+        self._assumed_state = host is None
 
     @property
     def is_on(self):
@@ -100,6 +97,16 @@ class WolSwitch(SwitchEntity):
     def name(self):
         """Return the name of the switch."""
         return self._name
+
+    @property
+    def assumed_state(self):
+        """Return true if no host is provided."""
+        return self._assumed_state
+
+    @property
+    def should_poll(self):
+        """Return false if assumed state is true."""
+        return not self._assumed_state
 
     def turn_on(self, **kwargs):
         """Turn the device on."""
@@ -118,38 +125,39 @@ class WolSwitch(SwitchEntity):
 
         wakeonlan.send_magic_packet(self._mac_address, **service_kwargs)
 
-        if self._is_dummy:
+        if self._assumed_state:
             self._state = True
+            self.async_write_ha_state()
 
     def turn_off(self, **kwargs):
         """Turn the device off if an off action is present."""
         if self._off_script is not None:
             self._off_script.run(context=self._context)
 
-        if self._is_dummy:
+        if self._assumed_state:
             self._state = False
+            self.async_write_ha_state()
 
     def update(self):
-        """Check if device is on and update the state."""
-        if not self._is_dummy:
-            if platform.system().lower() == "windows":
-                ping_cmd = [
-                    "ping",
-                    "-n",
-                    "1",
-                    "-w",
-                    str(DEFAULT_PING_TIMEOUT * 1000),
-                    str(self._host),
-                ]
-            else:
-                ping_cmd = [
-                    "ping",
-                    "-c",
-                    "1",
-                    "-W",
-                    str(DEFAULT_PING_TIMEOUT),
-                    str(self._host),
-                ]
+        """Check if device is on and update the state. Only called if assumed state is false."""
+        if platform.system().lower() == "windows":
+            ping_cmd = [
+                "ping",
+                "-n",
+                "1",
+                "-w",
+                str(DEFAULT_PING_TIMEOUT * 1000),
+                str(self._host),
+            ]
+        else:
+            ping_cmd = [
+                "ping",
+                "-c",
+                "1",
+                "-W",
+                str(DEFAULT_PING_TIMEOUT),
+                str(self._host),
+            ]
 
-            status = sp.call(ping_cmd, stdout=sp.DEVNULL, stderr=sp.DEVNULL)
-            self._state = not bool(status)
+        status = sp.call(ping_cmd, stdout=sp.DEVNULL, stderr=sp.DEVNULL)
+        self._state = not bool(status)

--- a/tests/components/wake_on_lan/test_switch.py
+++ b/tests/components/wake_on_lan/test_switch.py
@@ -275,3 +275,44 @@ async def test_invalid_hostname_windows(hass):
 
         state = hass.states.get("switch.wake_on_lan")
         assert STATE_OFF == state.state
+
+
+async def test_no_hostname_state(hass):
+    """Test that the state updates if we do not pass in a hostname."""
+
+    assert await async_setup_component(
+        hass,
+        switch.DOMAIN,
+        {
+            "switch": {
+                "platform": "wake_on_lan",
+                "mac": "00-01-02-03-04-05",
+            }
+        },
+    )
+    await hass.async_block_till_done()
+
+    state = hass.states.get("switch.wake_on_lan")
+    assert STATE_OFF == state.state
+
+    with patch.object(subprocess, "call", return_value=0):
+
+        await hass.services.async_call(
+            switch.DOMAIN,
+            SERVICE_TURN_ON,
+            {ATTR_ENTITY_ID: "switch.wake_on_lan"},
+            blocking=True,
+        )
+
+        state = hass.states.get("switch.wake_on_lan")
+        assert STATE_ON == state.state
+
+        await hass.services.async_call(
+            switch.DOMAIN,
+            SERVICE_TURN_OFF,
+            {ATTR_ENTITY_ID: "switch.wake_on_lan"},
+            blocking=True,
+        )
+
+        state = hass.states.get("switch.wake_on_lan")
+        assert STATE_OFF == state.state

--- a/tests/components/wake_on_lan/test_switch.py
+++ b/tests/components/wake_on_lan/test_switch.py
@@ -293,7 +293,7 @@ async def test_no_hostname_state(hass):
     await hass.async_block_till_done()
 
     state = hass.states.get("switch.wake_on_lan")
-    assert STATE_OFF == state.state
+    assert state.state == STATE_OFF
 
     with patch.object(subprocess, "call", return_value=0):
 
@@ -305,7 +305,7 @@ async def test_no_hostname_state(hass):
         )
 
         state = hass.states.get("switch.wake_on_lan")
-        assert STATE_ON == state.state
+        assert state.state == STATE_ON
 
         await hass.services.async_call(
             switch.DOMAIN,
@@ -315,4 +315,4 @@ async def test_no_hostname_state(hass):
         )
 
         state = hass.states.get("switch.wake_on_lan")
-        assert STATE_OFF == state.state
+        assert state.state == STATE_OFF

